java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java

```diff
@@ -90,169 +89,173 @@ private FlowComputation(@Nullable SymbolicValue symbolicValue, Predicate<Constra
-
-      Symbol newTrackSymbol = addFlowFromLearnedAssociations(path);
-      Stream<Constraint> learnedConstraints = addFlowFromLearnedConstraints(path.flow, path.edge());
-      if (parent.edges().isEmpty() || learnedConstraints.anyMatch(terminateTraversal)) {
+      ExecutionPath path = workList.pop();
+      if (path.finished) {
         flows.add(path.flow);
       } else {
-        parent.edges().stream()
-          .filter(e -> !path.contains(e))
-          .forEach(e -> workList.push(path.push(e, newTrackSymbol)));
+        path.lastEdge.parent.edges().stream()
+          .filter(path::notVisited)
+          .map(path::addEdge)
+          .forEach(workList::push);
       }
     }
     return flows;
   }
 
-  private Stream<Constraint> addFlowFromLearnedConstraints(List<JavaFileScannerContext.Location> flow, ExplodedGraph.Edge edge) {
-    Set<LearnedConstraint> learnedConstraints = edge.learnedConstraints();
-    Stream.Builder<Constraint> builder = Stream.builder();
-    for (LearnedConstraint learnedConstraint : learnedConstraints) {
-      if (symbolicValues.contains(learnedConstraint.symbolicValue()) && learnedConstraintForDomains(learnedConstraint)) {
-        Stream<JavaFileScannerContext.Location> locationStream = learnedConstraintFlow(learnedConstraint, edge);
-        locationStream.forEach(flow::add);
-        builder.add(learnedConstraint.constraint);
-      }
-    }
-    return builder.build();
-  }
-
-  private boolean learnedConstraintForDomains(LearnedConstraint lc) {
-    Constraint constraint = lc.constraint;
-    return constraint == null || domains.stream().anyMatch(d -> d.isAssignableFrom(constraint.getClass()));
-  }
+  private class ExecutionPath {
+    @Nullable final Symbol trackSymbol;
+    final ExplodedGraph.Edge lastEdge;
+    final PSet<ExplodedGraph.Edge> visited;
+    final List<JavaFileScannerContext.Location> flow;
+    final boolean finished;
 
-  private Stream<JavaFileScannerContext.Location> learnedConstraintFlow(LearnedConstraint learnedConstraint, ExplodedGraph.Edge edge) {
-    Constraint constraint = learnedConstraint.constraint();
-    if (constraint == null || !addToFlow.test(constraint)) {
-      return Stream.empty();
-    }
-    ExplodedGraph.Node parent = edge.parent;
-    Tree nodeTree = parent.programPoint.syntaxTree();
-    if (isMethodInvocationNode(parent)) {
-      return methodInvocationFlow(constraint, edge);
+    ExecutionPath(@Nullable ExplodedGraph.Edge edge, PSet<ExplodedGraph.Edge> visited, @Nullable Symbol trackSymbol, List<JavaFileScannerContext.Location> flow, boolean finished) {
+      this.trackSymbol = trackSymbol;
+      this.lastEdge = edge;
+      this.visited = visited;
+      this.flow = flow;
+      this.finished = finished;
     }
-    if (nodeTree.is(Tree.Kind.NEW_CLASS)) {
-      return Stream.of(location(parent, String.format("Constructor implies '%s'.", constraint.valueAsString())));
+
+    ExecutionPath(@Nullable Symbol trackSymbol) {
```

Comment:
I tend to think a "startPath(edge, trackSymbol)" method rather than this constructor would be nicer : so there would be no instanciation of this class from the outside.